Add extra Salt Bundle details (#1458)
uyuni-project/uyuni-docs: modules/client-configuration/pages/contact-methods-saltbundle.adoc

[[contact-methods-salt]]
= Salt Bundle

== What is Salt Bundle?

The Salt Bundle is a single binary package containing Salt Minion, Python 3, required Python modules and libraries.

> As the Salt Bundle is shipped with Python 3 and all the requirements for Salt. The Bundle is not using the Python installed on the system, so it can be installed on the client which doesn't meet the requirements for Salt version.

Reviewer: I noticed something weird and just rephrase this as:

```suggestion
As the Salt Bundle is shipped with Python 3 and all the requirements for Salt to run, the Salt Bundle is not using the Python version coming from the system, so it can be installed on clients which doesn't meet the requirements for a given Salt version.
```

> It is also possible to install Salt Bundle on the system where Salt Minion is being used already to manage that client by the different Salt Master than the {productname} Server.

Reviewer 2: ```suggestion
It is also possible to use the Salt Bundle on systems that run a Salt Minion connected to a non-{poductname} Salt Master.
```

Author: Not really sure if `non-Uyuni` or `non-SUSE Manager` cold look nice there.
Maybe do it this way:
```suggestion
It is also possible to use the Salt Bundle on systems that run a Salt Minion connected to a Salt Master other than {poductname} Salt Master.
```

Reviewer 2: You're right, `non-SUSE Manager` looks bad. I tried to avoid repeating "Salt Master", but I agree that your version is better than mine.

== Client Registration with Salt Bundle as a Minion

> The registration method with the Salt bundle is the recommended registration method as long as it covers your use case.

Reviewer 2: IMO that goes without saying.
```suggestion
The registration method with the Salt bundle is the recommended registration method.
```

Author: I think it was the suggestion by @meaksh from the initial docs PR for Salt Bundle. Personally, I like the original more.

Reviewer 2: If we both like the shorter one more, let's see what Pablo says :smiley: 
@meaksh how badly do you want to keep `as long as it covers your use case`?

Reviewer: I'm fine removing that. It was not actually me who wrote that. :smile:

This section explains the advantages and limitations of the current implementation.

> The Salt bundle is provided as the [package]``venv-salt-minion`` that consists of Salt, Python, and the Python modules Salt depends on.

Reviewer 2: Just for consistency with above
```suggestion
The Salt Bundle is provided as the [package]``venv-salt-minion`` that consists of Salt, Python 3, and the Python modules Salt depends on.
```

Bootstrapping with {webui} is using Salt bundle as well, so bootstrapping with {webui} is not Python dependant.

> Using the Salt bundle, it is no longer needed that the client provides the required Python version or any Python modules.

Reviewer 2: ```suggestion
Using the Salt Bundle, it is no longer needed that the client provides any Python interpreter or modules.
```

If you bootstrap new clients, registration with the Salt bundle is the default method.
You can switch existing clients to the Salt bundle method.
If you switch, the [package]``salt-minion`` package and its Python version and the Python modules will stay installed.

=== Switching from Salt Minion to Salt Bundle

The Salt state [literal]``util.mgr_switch_to_venv_minion`` is available to switch from [package]``salt-minion`` to [package]``venv-salt-minion``.

> It is recommended to switch to [package]``venv-salt-minion`` in two steps to avoid trouble with shifting processes:

Author: This part was there since the initial PR with Salt Bundle documentation: https://www.uyuni-project.org/uyuni-docs/en/uyuni/client-configuration/registration-methods-saltbundle.html
Just in the other section. I haven't touched this part.

> It is recommended to switch to [package]``venv-salt-minion`` in two steps to avoid trouble with shifting processes:

Author: Slightly rephased it and added extra section with the warning that the state apply could fail in case of running the second step with skipping the first.

.Procedure: Switching with [literal]``util.mgr_switch_to_venv_minion`` state to [package]``venv-salt-minion``

. Apply [literal]``util.mgr_switch_to_venv_minion`` with no pillar specified first.
  This will result in the switch to [package]``venv-salt-minion`` with copying configuration files etc.
  It will not clean up the original [package]``salt-minion`` configurations and its packages.
+
----
salt <minion_id> state.apply util.mgr_switch_to_venv_minion
----
. Apply [literal]``util.mgr_switch_to_venv_minion`` with [literal]``mgr_purge_non_venv_salt`` set to [literal]``True`` to remove [package]``salt-minion`` and with [literal]``mgr_purge_non_venv_salt_files`` set to [literal]``True`` to remove all the files related to [package]``salt-minion``.

> This second step repeats the first step, and then removes the old configuration files and the now obsolete [package]``salt-minion`` package.

Author: This can't be done in 1 step as it requires starting new minion service and stopping the old one. In general the second step is not repeating, but checking it the steps from first step were performed. Maybe need to rephrase it somehow.
```
The second step ensures that the first step passed and no changes required for it ....
```

+
----
salt <minion_id> state.apply util.mgr_switch_to_venv_minion pillar='{"mgr_purge_non_venv_salt_files": True, "mgr_purge_non_venv_salt": True}'
----

On the other hand, it is also possible to avoid installing the Salt bundle and keep using [package]``salt-minion`` instead.

> In this case specify one of these options:

Reviewer 2: ```suggestion
In this case, specify one of these options:
```

* Execute [command]``mgr-bootstrap`` with  [option]``--no-bundle`` option.
* Set [literal]``AVOID_VENV_SALT_MINION`` to [literal]``1`` in the generated bootstrap script.
* For bootstrap state set the [literal]``mgr_avoid_venv_salt_minion`` pillar to  [literal]``True``.

== Salt SSH with the Salt Bundle

The Salt Bundle is also used to handle Salt SSH session on the client side.

> The pre flight script is deploying the Salt Bundle on the client without installing the package. Original `salt-thin` method is not used in this case.

Reviewer: Another suggestion: 

```suggestion
A pre flight script is executed as part of the Salt SSH execution which deploys the code from the Salt Bundle on the target client without installing the package. The original `salt-thin` method is not used in this case.
```

Maybe we could also mention here the paths where bundle gets deployed. Not sure.

Author: Not sure if it makes any sense to the end users to know where it is placed, in general it's quite easy to find if they are really interested.

> A pre flight script is executed as part of the Salt SSH execution which deploys the code from the Salt Bundle on the target client without installing the package. The original `salt-thin` method is not used in this case.

Reviewer 2: I would join this paragraph with the one below and reword it slightly.
```suggestion
A shell script deploys the Salt Bundle onto the target system without installing [package]``venv-salt-minion`` before any Salt command is executed. Since the Salt Bundle contains the whole Salt code base, no [literal]``salt-thin`` is deployed. Salt SSH (including bootstrapping using the {webui}) uses the Python 3 interpreter within the bundle. The target system does not need to have any other Python interpreter installed.
```

> Salt API is calling Python 3 deployed with the Bundle to handle Salt SSH session on the client, so Salt SSH (including bootstrapping with {webui}) is not dependant on Python installed on the system.

Author: We can avoid mentioning Salt API for example the following way:
```
The Python 3 deployed with the Bundle is used to handle Salt SSH session on the client, so Salt SSH (including bootstrapping with {webui}) is not dependant on Python installed on the system.
```
Is it better?

[IMPORTANT]
====

> The bootstrap repository must be created before bootstrapping the client with {webui}. The pre flight script is detecting the OS on the client and using the bootstrap repository relevant for the OS version.

Reviewer 2: I think we don't need to expose the "pre flight script" name to our users, I see it only as an implementation detail.
```suggestion
The bootstrap repository must be created before bootstrapping the client with {webui}. The Salt Bundle used by Salt SSH is taken from the bootstrap repository based on the detected target operating system.
```

Author: Not sure if it sounds more clear in passive than before.

Reviewer: I'm also in favor of not mentioning pre flight script, maybe we can put the sentence in a different way but without mentioning pre flight script.

For more information, see xref:client-configuration:bootstrap-repository.adoc#_prepare_to_create_a_bootstrap_repository[Prepare to Create a Bootstrap Repository]
====

[NOTE]
====
Using `salt-thin` can be enabled as a fallback method, but it requires Python 3 to be installed on the client.

> This method is not recommended and supported and can be used for development purposes only.

Reviewer 2: ```suggestion
This method is not recommended nor supported and exists for development purposes only.
```

Set [option]``web.ssh_use_salt_thin`` to `true` in the [path]``/etc/rhn/rhn.conf`` configuration file.
====

== Using the Salt Bundle with the Salt Minion

The Salt Bundle can be used with the Salt Minion managed by the Salt Master other than {productname} Server at the same time.

[IMPORTANT]
====

> To bootstrap the client with the Salt Minion managed by the Salt Master other than {productname} Server it is recommended to use [command]``mgr-bootstrap`` [option]``--force-bundle`` on generating bootstrap script, or set [option]``FORCE_VENV_SALT_MINION`` to `1` in the bootstrap script.

Author: I'll fix it just a bit different way.
